### PR TITLE
Adding OTA agent files

### PR DESCRIPTION
**Pull request by divekarshubham**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by tgsong — CHANGES_REQUESTED** [review]

**Review by tgsong — CHANGES_REQUESTED** [review]

File: source/ota.c [hunk 1]
```diff
@@ -0,0 +1,3028 @@
...
+                LogInfo( ( "Job was accepted. Attempting to start transfer.\r\n" ) );
+            }
+        }
+    }
+    else
+    {
+        LogError( ( "Error %d parsing job document.\r\n", err ) );
+    }
+
+    if( setError != OTA_ERR_NONE )
```
[the remark this answers is not in the document]
Comment (divekarshubham) [reply source/ota.c]: Removed in https://github.com/aws/ota-for-aws-iot-embedded-sdk/commit/b29e98996f388af68e3a033ad7464cd78f77716a

File: source/ota.c [hunk 2]
```diff
@@ -0,0 +1,3028 @@
...
+OtaErr_t OTA_SetImageState( OtaImageState_t state )
+{
+    OtaErr_t err = OTA_ERR_UNINITIALIZED;
+    OtaEventMsg_t eventMsg = { 0 };
+
+    switch( state )
+    {
+        case OtaImageStateAborted:
+
+            if( 1 /*otaAgent.xOTA_EventQueue != NULL*/ )
```
[the remark this answers is not in the document]
Comment (pvyawaha) [reply source/ota.c]: This is fixed

File: source/include/ota.h [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+ * @file ota.h
+ * @brief OTA Agent Interface
+ */
+
+#ifndef _AWS_IOT_OTA_AGENT_H_
+#define _AWS_IOT_OTA_AGENT_H_
+
+/* Standard includes. */
+/* For FILE type in OtaFileContext_t.*/
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (pvyawaha) [reply source/include/ota.h:36]: I think we should use the file ptr for windows/linux

File: source/include/ota.h [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+ * @brief OTA File Context Information.
+ *
+ * Information about an OTA Update file that is to be streamed. This structure is filled in from a
+ * job notification MQTT message. Currently only one file context can be streamed at time.
+ */
+struct OtaFileContext
+{
+    uint8_t * pFilePath;      /*!< Local file pathname. */
+    #if defined( WIN32 ) || defined( __linux__ )
+        FILE * pFile;         /*!< File type is stdio FILE structure after file is open for write. */
```
[the remark this answers is not in the document]
Comment (pvyawaha) [reply source/include/ota.h]: See my comment above

File: source/include/ota_private.h [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+ *       to attain the effective address (somewhere in RAM). Otherwise, it is interpreted
+ *       as an absolute memory address and used as is (useful for singleton parameters).
+ *       This requires absolute memory pointers to be greater than the size of the
+ *       context structure to avoid the address being misinterpreted as an offset.
+ */
+typedef struct
+{
+    const char * pSrcKey;                  /*!< Expected key name. */
+    const bool required;                   /*!< If true, this parameter must exist in the document. */
+    void * const pDestOffset;              /*!< Pointer or offset to where we will store the value, if not ~0. */
```
[the remark this answers is not in the document]
Comment (pvyawaha) [reply source/include/ota_private.h]: Updated.

File: source/ota.c [hunk 6]
```diff
@@ -0,0 +1,3173 @@
...
+    .serverFileID               = 0,
+    .pOtaSingletonActiveJobName = NULL,
+    .pClientTokenFromJob        = NULL,
+    .timestampFromJob           = 0,
+    .imageState                 = OtaImageStateUnknown,
+    .palCallbacks               = OTA_JOB_CALLBACK_DEFAULT_INITIALIZER,
+    .numOfBlocksToReceive       = 1,
+    .statistics                 = { 0 },
+    .requestMomentum            = 0
+};
```
[the remark this answers is not in the document]
Comment (pvyawaha) [reply source/ota.c:298]: Updated.

File: source/ota.c [hunk 7]
```diff
@@ -0,0 +1,3183 @@
...
+
+        result = true;
+    }
+
+    return result;
+}
+
+/* Find an available OTA transfer context structure. */
+
+static OtaFileContext_t * getFreeContext( void )
```
[the remark this answers is not in the document]
Comment (pvyawaha) [reply source/ota.c:1303]: Removed.

File: source/ota.c [hunk 8]
```diff
@@ -0,0 +1,3028 @@
...
+    /* If we're here, we're shutting down the OTA agent. Free up all resources and quit. */
+    agentShutdownCleanup();
+
+    /* Clear the entire agent context. This includes the OTA agent state. */
+    ( void ) memset( &otaAgent, 0, sizeof( otaAgent ) );
+
+    otaAgent.state = OtaAgentStateStopped;
+
+    /* Terminate the OTA Agent Thread. */
+    pthread_exit( NULL );
```
[the remark this answers is not in the document]
Comment (tgsong) [reply source/ota.c:1210]: it's still here, on both this pr and development branch.

File: source/include/ota.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgsong) [reply source/include/ota.h:36]: I remember MISRA prohibits the use of `stdio.h` header. Why we need this to be a concrete type? Only the pal cares about this pointer, and they can convert it right?

File: source/ota.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (tgsong) [reply source/ota.c:296]: Where is this updated? It's not showing here.

File: source/include/ota.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (pvyawaha) [reply source/include/ota.h:36]: yes I can note it and we can do a separate PR as we will have to update the PALs for linux and windows as well.

File: source/ota.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (divekarshubham) [reply source/ota.c:296]: This is moved inside pOtaInterface, can be found in https://github.com/aws/ota-for-aws-iot-embedded-sdk/pull/52/commits/6db0fb222600b968b2c8cb037c61577c4aab2a2c

File: source/ota.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (divekarshubham) [reply source/ota.c]: Removed in https://github.com/aws/ota-for-aws-iot-embedded-sdk/commit/5f3ea308664d2f0c0de1a7798a9551f2fba6e4f4

File: source/ota.c [hunk 9]
```diff
@@ -0,0 +1,3028 @@
...
+    {
+        err = decodeAndStoreKey( pValueInJson, valueLength, ppvParamAdd );
+    }
+    else if( ModelParamTypeIdent == docParam.modelParamType )
+    {
+        LogDebug( ( "Identified parameter [ %s ]\r\n",
+
+                    docParam.pSrcKey ) );
+
+        *ppvParamAdd = true;
```
[the remark this answers is not in the document]
Comment (divekarshubham) [reply source/ota.c]: updated in https://github.com/aws/ota-for-aws-iot-embedded-sdk/commit/5f3ea308664d2f0c0de1a7798a9551f2fba6e4f4

File: source/ota.c [hunk 10]
```diff
@@ -0,0 +1,3089 @@
...
+    {
+        parseError = parseJSONbyModel( pJson, messageLength, &otaJobDocModel );
+
+        if( parseError == DocParseErrNone )
+        {
+            err = validateAndStartJob( pFileContext, &pFinalFile, pUpdateJob );
+        }
+        else
+        {
+            err = parseJobDocFromCustomCallback( pJson, messageLength, pFileContext );
```
[the remark this answers is not in the document]
Comment (divekarshubham) [reply source/ota.c]: fixed in https://github.com/aws/ota-for-aws-iot-embedded-sdk/commit/5f3ea308664d2f0c0de1a7798a9551f2fba6e4f4

**Review by pvyawaha — DISMISSED** [review]

**Review by tgsong — DISMISSED** [review]

**Review by tgsong — DISMISSED** [review]

**Review by tgsong — APPROVED** [review]

**Review by n9wxu — APPROVED** [review]

**Pull request merged**